src/Orleans.Runtime/GrainDirectory/GrainLocatorResolver.cs

```diff
@@ -0,0 +1,50 @@
+using System;
+using System.Collections.Concurrent;
+using Orleans.GrainDirectory;
+
+namespace Orleans.Runtime.GrainDirectory
+{
+    internal class GrainLocatorResolver
+    {
+        private readonly ConcurrentDictionary<GrainType, IGrainLocator> resolvedLocators = new ConcurrentDictionary<GrainType, IGrainLocator>(GrainType.Comparer.Instance);
+        private readonly Func<GrainType, IGrainLocator> getLocatorInternal;
+        private readonly GrainDirectoryResolver grainDirectoryResolver;
+        private readonly CachedGrainLocator cachedGrainLocator;
+        private readonly DhtGrainLocator dhtGrainLocator;
+        private readonly ClientGrainLocator clientGrainLocator;
+
+        public GrainLocatorResolver(
+            GrainDirectoryResolver grainDirectoryResolver,
+            CachedGrainLocator cachedGrainLocator,
+            DhtGrainLocator dhtGrainLocator,
+            ClientGrainLocator clientGrainLocator)
+        {
+            this.getLocatorInternal = GetGrainLocatorInternal;
+            this.grainDirectoryResolver = grainDirectoryResolver;
+            this.cachedGrainLocator = cachedGrainLocator;
+            this.dhtGrainLocator = dhtGrainLocator;
+            this.clientGrainLocator = clientGrainLocator;
+        }
+
+        public IGrainLocator GetGrainLocator(GrainType grainType) => resolvedLocators.GetOrAdd(grainType, this.getLocatorInternal);
+
+        public IGrainLocator GetGrainLocatorInternal(GrainType grainType)
+        {
+            IGrainLocator result;
+            if (grainType.IsClient())
+            {
+                result = this.clientGrainLocator;
+            }
+            else if (this.grainDirectoryResolver.HasNonDefaultDirectory(grainType))
+            {
+                result = this.cachedGrainLocator;
+            }
+            else
+            {
+                result = this.dhtGrainLocator;
```

Comment:
What if the default is not set to the `dhtGrainLocator`?

Reply:
I think this is the same behavior as it is today. Either way, we can merge now & follow up